### PR TITLE
ARROW-4845: [R] Compiler warnings on Windows MingW64

### PR DESCRIPTION
**Pull request by romainfrancois**

### CODE REVIEW

**Comment by romainfrancois** [comment]
This did not test on win builder because I supposed we need a new build of `libarrow`, #4413 followed up on #4316 and introduced some incompatibilities. https://win-builder.r-project.org/KG75rToaX3Qd/

How do we trigger new builds for https://github.com/rwinlib/arrow ? @jeroen @javierluraschi

**Comment by nealrichardson** [comment]
https://issues.apache.org/jira/browse/ARROW-4845 shows lots of warnings that look to be from different lines--are you sure this is the only change needed?

**Comment by romainfrancois** [comment]
I'm not sure, that's why I tried to trigger a win builder build with `devtools::check_win_devel()` but it ended with this: 

```
*** arch - i386
"D:/RCompile/recent/R/bin/i386/Rscript.exe" "../tools/winlibs.R" 0.13.0.9000
Error in download.file(sprintf("https://github.com/rwinlib/arrow/archive/v%s.zip",  : 
  cannot open URL 'https://github.com/rwinlib/arrow/archive/v0.13.0.9000.zip'
In addition: Warning message:
In download.file(sprintf("https://github.com/rwinlib/arrow/archive/v%s.zip",  :
  cannot open URL 'https://github.com/rwinlib/arrow/archive/v0.13.0.9000.zip': HTTP status was '404 Not Found'
Execution halted
make: *** [winlibs] Error 1
ERROR: compilation failed for package 'arrow'
* removing 'd:/RCompile/CRANguest/R-devel/lib/arrow'
```

before it could show me compile errors.

File: r/src/recordbatch.cpp [hunk 1]
```diff
@@ -131,7 +131,7 @@ std::shared_ptr<arrow::RecordBatch> ipc___ReadRecordBatch__InputStream__Schema(
 arrow::Status check_consistent_array_size(
     const std::vector<std::shared_ptr<arrow::Array>>& arrays, int64_t* num_rows) {
   *num_rows = arrays[0]->length();
-  for (int64_t i = 1; i < arrays.size(); i++) {
+  for (size_t i = 1; i < arrays.size(); i++) {
```
Comment (pitrou) [inline r/src/recordbatch.cpp]: Idiomatic C++11 style would be something like this:
```cpp
for (const auto& array : arrays) {
  if (array->length() != *num_rows) {
    // etc.
```

Which also avoids having to think about index type.

File: r/src/recordbatch.cpp [hunk 2]
```diff
@@ -131,7 +131,7 @@ std::shared_ptr<arrow::RecordBatch> ipc___ReadRecordBatch__InputStream__Schema(
 arrow::Status check_consistent_array_size(
     const std::vector<std::shared_ptr<arrow::Array>>& arrays, int64_t* num_rows) {
   *num_rows = arrays[0]->length();
```
Comment (pitrou) [inline r/src/recordbatch.cpp]: What happens if there are 0 arrays?

**Review by pitrou — COMMENTED** [review]
Looks like the fix would indeed silence the warning, but some comments.

**Comment by romainfrancois** [comment]
Perhaps we can use `-Wall` on travis to get some other warnings as well.

File: r/src/recordbatch.cpp (hunk 1 above)
Comment (romainfrancois) [reply r/src/recordbatch.cpp]: Thanks.

File: r/src/recordbatch.cpp (hunk 2 above)
Comment (romainfrancois) [reply r/src/recordbatch.cpp]: 😬 thanks, updated.

**Comment by jeroen** [comment]
Once libarrow 0.14 is released you can send a PR to https://github.com/r-windows/rtools-packages and https://github.com/r-windows/rtools-backports and we can upload new binaries on rwinlib.

**Comment by wesm** [comment]
At some point we need to document the source build process on Windows rather than relying on pre-packaged binaries so that we can run it in CI, otherwise we will eventually get slammed with a Windows-only problem post-release

https://issues.apache.org/jira/browse/ARROW-3758

**Comment by jeroen** [comment]
The build process is [documented](https://github.com/r-windows/rtools-packages#building-package) and [automated](https://ci.appveyor.com/project/jeroen/rtools-backports/history). 

You can either build libarrow locally or send a PR to https://github.com/r-windows/rtools-backports and then download the binaries from appveyor, and build your R package with that.

There is nothing special about [rwinlib](https://github.com/rwinlib), it's just a place where we host binaries for (stable) libraries. But you could host them anywhere else.

**Pull request closed without merging**